### PR TITLE
Psql selection

### PR DESCRIPTION
**Pull request by edpaget**
Not super happy with some bits of this (mostly how it does the retired/already seen marking), but I think it should work for Stargazing.

### CODE REVIEW

File: lib/postgresql_selection.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
+class PostgresqlSelection
+  attr_reader :workflow, :user
+  
+  def initialize(workflow, user=nil)
+    @workflow, @user = workflow, user
+  end
+
+  def select(opts={})
+    limit = opts.fetch(:limit, 20).to_i
+    selection_statement = case
```
Comment (camallen) [inline lib/postgresql_selection.rb]: can you extract this to a method?

File: app/controllers/api/v1/subjects_controller.rb [hunk 2]
```diff
@@ -21,6 +24,11 @@ def index
 
   private
 
+  def selector_context
+    { retired: workflow.finished?,
```
Comment (camallen) [inline app/controllers/api/v1/subjects_controller.rb]: Is this being tested anywhere in the subjects_controller responses?

Comment (edpaget) [inline app/controllers/api/v1/subjects_controller.rb]: I added some tests.

**Comment by camallen** [comment]
Looks fine to merge.

In benchmarking mood and am wondering what sort of performance hit this might add to selection routes.

**Comment by edpaget** [comment]
Yeah loading the workflow and doing the `has_finished?` and `finished?` methods will add some overhead. I think it could be avoided with some changes to the subject selector, but I'm not exactly sure.

File: app/controllers/api/v1/subjects_controller.rb [hunk 3]
```diff
@@ -11,9 +11,12 @@ class Api::V1::SubjectsController < Api::ApiController
   def index
     case params[:sort]
     when 'cellect'
-      render json_api: SubjectSerializer.page(params, selector.cellect_subjects)
+      render json_api: SubjectSerializer.page(params,
+                                              selector.cellect_subjects,
+                                              selector_context)
```
Comment (camallen) [inline app/controllers/api/v1/subjects_controller.rb]: Just had a thought re avoiding paying the cost with context methods. Can we move the `selector_context` into the selector object and switch it depending on whether cellect returned results? It's be nice not to check if the workflow is finished / a user has finished each time we want subjects.

**Pull request merged**